File: Curso Java Com POO/Java com POO/Aula09(Exercicio Pratico)/src/teste/Livro.java
package teste;

//Classe Livro agregação com a Classe Pessoa
//Classe Livro implementa a interface Publicacao
public class Livro implements Publicacao {
    //Atributos
    private String titulo;
    private String autor;
    private int totPaginas;
    private int pagAtual;
    private boolean aberto;
    private Pessoa leitor; //atributo de agregação
    
    //Métodos
    //usei o método "toString()..." em alt+insert 
    public String detalhes() {
        return "Livro{" 
                + "\n titulo=" + titulo + ", autor=" + autor 
                + ",\n totPaginas=" + totPaginas + ", pagAtual=" + pagAtual + ", aberto=" + aberto 
                + ",\n leitor=" + leitor.getNome() + ", idade=" + leitor.getIdade() + ", sexo=" + leitor.getSexo()
                + '}';
    }
    
    //Construtor
    public Livro(String titulo, String autor, int totPaginas, Pessoa leitor) {
        this.titulo = titulo;
        this.autor = autor;
        this.totPaginas = totPaginas;
        this.aberto = false;
        this.pagAtual = 0;
        this.leitor = leitor;
    }
    
    //Getters e Setters
    //titulo
    public String getTitulo() {
        return titulo;
    }
    public void setTitulo(String titulo) {
        this.titulo = titulo;
    }

    //autor
    public String getAutor() {
        return autor;
    }
    public void setAutor(String autor) {
        this.autor = autor;
    }

    //totPaginas
    public int getTotPaginas() {
        return totPaginas;
    }
    public void setTotPaginas(int totPaginas) {
        this.totPaginas = totPaginas;
    }

    //pagAtual
    public int getPagAtual() {
        return pagAtual;
    }
    public void setPagAtual(int pagAtual) {
        this.pagAtual = pagAtual;
    }

    //aberto
    public boolean getAberto() {
        return aberto;
    }
    public void setAberto(boolean aberto) {
        this.aberto = aberto;
    }

    //leitor
    public Pessoa getLeitor() {
        return leitor;
    }
    public void setLeitor(Pessoa leitor) {
        this.leitor = leitor;
    }
    
    //Métodos da interface Publicacao
    @Override
    public void abrir() {
        this.setAberto(true);
        this.setPagAtual(1);
    }

    @Override
    public void fechar() {
        this.setAberto(false);
        this.setPagAtual(0);
    }

    @Override
    public void folhear(int p) {
        if (this.getAberto()) {
            if (p > this.totPaginas) {
                this.pagAtual = 1;
            } else {
                this.setPagAtual(p);
            } 
        } else {
            System.out.println("Não consigo folhear! O livro está fechado.");
        }   
    }

    @Override
    public void avancarPag() {
        if (this.getAberto()) {
            if (this.getPagAtual() == this.getTotPaginas()) {
                System.out.println("Não consigo avançar! Você está na última página.");
            } else {
                this.setPagAtual(this.getPagAtual() + 1);
                //this.pagAtual++; -> outra opção
            }
        } else {
            System.out.println("Não consigo avançar! O livro está fechado.");
        }             
    }

    @Override
    public void voltarPag() {
        if (this.getAberto()) {
            if (this.getPagAtual() == 0) {
                System.out.println("Não consigo voltar! Você está na primeira página.");
            } else {
                this.setPagAtual(this.getPagAtual() - 1);
                //this.pagAtual--; -> outra opção
            }
        } else {
            System.out.println("Não consigo voltar! O livro está fechado.");
        }             
    }
}

File: Curso Programação Orientada a Objetos/PHP com POO/Aula09-ExercicioPratico/Livro.php
<?php

require_once 'Pessoa.php';
require_once 'Publicacao.php';

class Livro implements Publicacao {
    //Atributos
    private $titulo;
    private $autor;
    private $totPaginas;
    private $pagAtual;
    private $aberto;
    private $leitor; 
    
    //Métodos
    public function detalhes() {
        return "Livro"
                . "<br>[titulo=" . $this->titulo 
                . ", autor=" . $this->autor
                . ", <br> totPaginas=" . $this->totPaginas
                . ", pagAtual=" . $this->pagAtual
                . ", aberto=" . $this->aberto
                . ", <br> leitor=" . $this->leitor->getNome()
                . ", idade=" . $this->leitor->getIdade()
                . ", sexo=" . $this->leitor->getSexo()
                . "]";
    }
    
    //Construtor
    public function __construct($titulo, $autor, $totPaginas, $leitor) {
        $this->titulo = $titulo;
        $this->autor = $autor;
        $this->totPaginas = $totPaginas;
        $this->pagAtual = 0;
        $this->aberto = false;
        $this->leitor = $leitor;
    }
    
    //Getters
    public function getTitulo() {
        return $this->titulo;
    }

    public function getAutor() {
        return $this->autor;
    }

    public function getTotPaginas() {
        return $this->totPaginas;
    }

    public function getPagAtual() {
        return $this->pagAtual;
    }

    public function getAberto() {
        return $this->aberto;
    }

    public function getLeitor() {
        return $this->leitor;
    }
    
    
    //Setters
    public function setTitulo($titulo) {
        $this->titulo = $titulo;
    }

    public function setAutor($autor) {
        $this->autor = $autor;
    }

    public function setTotPaginas($totPaginas) {
        $this->totPaginas = $totPaginas;
    }

    public function setPagAtual($pagAtual) {
        $this->pagAtual = $pagAtual;
    }

    public function setAberto($aberto) {
        $this->aberto = $aberto;
    }

    public function setLeitor($leitor) {
        $this->leitor = $leitor;
    }

    
    //Métodos da interface Publicacao
    public function abrir() {
        $this->setAberto(true);
        $this->setPagAtual(1);
    }

    public function fechar() {
        $this->setAberto(false);
        $this->setPagAtual(0);
    }

    public function folhear($p) {
        if ($this->getAberto()) {
            if ($p > $this->totPaginas) {
                $this->pagAtual = 1;
            } else {
                $this->setPagAtual($p);
            }
        } else {
            echo "<p>Não consigo folhear! O livro está fechado.</p>";
        }
    }

    public function avancarPag() {
        if ($this->getAberto()) {
            if ($this->getPagAtual() === $this->getTotPaginas()) {
                echo "<p>Não consigo avançar! Você está na última página. </p>";
            } else {
                $this->setPagAtual($this->getPagAtual() + 1);
            }
        } else {
            echo "<p>Não consigo avançar! O livro está fechado.</p>";
        }
    }

    public function voltarPag() {
       if ($this->getAberto()) {
            if ($this->getPagAtual() === 0) {
                echo "<p>Não consigo voltar! Você está na primeira página. </p>";
            } else {
                $this->setPagAtual($this->getPagAtual() - 1);
            }
        } else {
            echo "<p>Não consigo voltar! O livro está fechado.</p>";
        }
    }   
}
